c10/core/impl/cow/state_machine.cpp

```diff
@@ -0,0 +1,240 @@
+        }
+        return *impl;
+      }
+
+      case StateId::initializing:
+        seen_initializing = true;
+        // Wait for initialization.
+        continue;
+
+      case StateId::active:
+        return &state_rep.as_state_machine_impl();
+    }
+  } while (true);
+}
+
+cow::StateMachine::Rep::Rep(cow::StateMachine::StateId state_id)
+    : rep_(static_cast<std::uintptr_t>(state_id)) {
+  TORCH_INTERNAL_ASSERT(as_state_id() == state_id);
+}
+
+namespace {
+
+// Implementations of std::bit_cast() from C++ 20.
+//
+// This is a less sketchy version of reinterpret_cast.
+//
+// See https://en.cppreference.com/w/cpp/numeric/bit_cast for more
+// information as well as the source of our implementations.
+template <class To, class From>
+std::enable_if_t<
+    sizeof(To) == sizeof(From) && std::is_trivially_copyable_v<From> &&
+        std::is_trivially_copyable_v<To>,
+    To>
+// constexpr support needs compiler magic
+bit_cast(const From& src) noexcept {
+  static_assert(
+      std::is_trivially_constructible_v<To>,
+      "This implementation additionally requires "
+      "destination type to be trivially constructible");
+
+  To dst;
+  std::memcpy(&dst, &src, sizeof(To));
+  return dst;
+}
+
+} // namespace
+
+cow::StateMachine::Rep::Rep(Impl& impl)
+    : rep_(bit_cast<std::uintptr_t>(&impl)) {
+  TORCH_INTERNAL_ASSERT(as_state_id() == StateId::active);
+}
+
+auto cow::StateMachine::Rep::as_state_id() const noexcept
+    -> cow::StateMachine::StateId {
+  return StateId{static_cast<int>(rep_ & 0b111)};
+}
+
+auto cow::StateMachine::Rep::as_state_machine_impl()
+    -> cow::StateMachine::Impl& {
+  TORCH_INTERNAL_ASSERT(as_state_id() == StateId::active);
+  TORCH_INTERNAL_ASSERT((rep_ & 0b111) == 0); // no alignment bits are modified
+  return *bit_cast<cow::StateMachine::Impl*>(rep_);
+}
+
+auto cow::StateMachine::Rep::operator==(const Rep& that) const noexcept
+    -> bool {
+  return rep_ == that.rep_;
+}
+
+namespace {
+
+// Applies a function to whichever shadow storage is active.
+//
+// Note that we templatize on the shadow storage types because they
+// may or may not be const. The bare types will always be
+// ShadowStorage and ShadowStorageNonIntrusive.
+template <typename ShadowStorage, typename DefaultShadowStorage, typename Func>
+auto apply_to_active_shadow_storage(
+    ShadowStorage* shadow_storage,
+    DefaultShadowStorage& default_shadow_storage,
+    const Func& func) {
+  if (shadow_storage != nullptr) {
+    return func(*shadow_storage);
+  }
+  return func(default_shadow_storage);
+}
+
+} // namespace
+
+auto cow::StateMachine::Impl::maybe_bump(
+    cow::ShadowStorage* maybe_shadow_storage) -> void {
+  std::lock_guard<std::mutex> lock(mtx_);
+  Generation physical_generation = ++physical_generation_;
+  Generation shadow_generation = apply_to_active_shadow_storage(
+      maybe_shadow_storage, default_shadow_storage_, [](auto& shadow_storage) {
+        return shadow_storage.bump_generation();
```

Comment:
It's a little weird to do the shadow storage as a bump, rather than assigning physical generation to it (and doing the warn check before you do the assignment.

The difference here is what happens if you do:

```
y = x.reshape(...)
x.add_(2)
y.add_(2)
```

The way you have this implemented, after the two mutations, both x and y's shadow storage are "one generation short" of physical generation, so everyone is going to complain on subsequent reads! But it seems more humane to warn on the y.add_(2) (because writes count as reads) and then just assume that y is "up to date" and any reads on it are OK.

Reply:
nice. done.